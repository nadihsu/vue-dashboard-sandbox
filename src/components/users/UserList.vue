<template lang="pug">
div
  el-table(:data="data", stripe, border)
    el-table-column(prop="id", label="序")
    el-table-column(prop="username", label="使用者")
    el-table-column(prop="created_at", label="建立時間", :formatter="formatterDate")

    el-table-column(prop="enable", label="啟用", align="center")
      template(#default="scope")
        el-tag(:type="scope.row.enable ? 'success' : 'danger'").
          {{ scope.row.enable ? '啟用' : '停用' }}

    el-table-column(prop="locked", label="鎖定", align="center")
      template(#default="scope2")
        el-icon(:size="20")
          i-ep-lock(v-if="scope2.row.locked", color="#999")
          i-ep-unlock(v-else).

    el-table-column(label="操作", align="center")
      template(#default="scope3")
        el-button(size="small", @click="openEditUserDialog(scope3.row)").
          編輯
        el-button(type="danger", size="small", @click="openDeleteUserDialog(scope3.row)").
          刪除

el-dialog(
v-model="showEditModal",
:title="`編輯使用者: ${user.username}`",
width="25%",
destroy-on-close)
  edit-user-modal(
  :active="showEditModal",
  :user="user",
  :close-modal="onClose",
  :edit-user="editUser").

el-dialog(
v-model="showDeleteModal",
:title="`刪除使用者: ${user.username}`",
width="25%",
center,
destroy-on-close)
  delete-user-modal(
  :active="showDeleteModal",
  :user="user",
  :close-modal="onClose",
  :delete-user="deleteUser").

</template>

<script>
import { ref, getCurrentInstance } from 'vue';
import EditUserModal from './EditUserModal.vue';
import DeleteUserModal from './DeleteUserModal.vue';

export default {
  components: { EditUserModal, DeleteUserModal },
  props: {
    data: {
      type: Array,
      required: true,
    },
    editUser: {
      type: Function,
      default: () => {},
    },
    deleteUser: {
      type: Function,
      default: () => {},
    },
  },
  setup() {
    const { proxy } = getCurrentInstance();
    const showEditModal = ref(false);
    const showDeleteModal = ref(false);
    const user = ref({});

    /**
     * 日期格式化
     *
     * @param {{ created_at: string }} row 使用者建立日期
     */
    function formatterDate(row) {
      return proxy.$day(row.created_at).format('YYYY-MM-DD');
    }

    /**
     * 關閉彈窗
     */
    function onClose() {
      showEditModal.value = false;
      showDeleteModal.value = false;
    }

    /**
     * 開啟編輯彈窗
     *
     * @param {object} userData
     */
    function openEditUserDialog(userData) {
      showEditModal.value = true;
      user.value = userData;
    }

    /**
     * 開啟刪除彈窗
     *
     * @param {object} userData
     */
    function openDeleteUserDialog(userData) {
      showDeleteModal.value = true;
      user.value = userData;
    }

    return {
      showEditModal,
      showDeleteModal,
      user,
      formatterDate,
      onClose,
      openEditUserDialog,
      openDeleteUserDialog,
    };
  },
};
</script>
