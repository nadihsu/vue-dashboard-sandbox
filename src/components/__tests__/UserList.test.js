import { render } from 'Tests/unit/TestUtils';
import UserList from '../users/UserList.vue';

describe('User Lists Page', () => {
  it('基本', async () => {
    const setup = render(
      UserList,
      {
        globalConfig: {
          stubs: ['EditUserModal', 'DeleteUserModal'],
        },
        props: {
          data: [
            {
              username: 'user1',
              enable: 1,
              locked: 0,
              created_at: '2023-02-08T14:00:00+08:00',
            },
          ],
        },
      },
    );

    const { findByText, getByText } = setup;

    await findByText('user1');

    expect(getByText('2023-02-08')).toBeInTheDocument();
    expect(setup.html()).toMatchSnapshot();
  });
});
