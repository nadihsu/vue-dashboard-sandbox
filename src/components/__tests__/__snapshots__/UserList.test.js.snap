// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`User Lists Page > 基本 1`] = `
<div>
  <div class="el-table--fit el-table--striped el-table--border el-table--enable-row-hover el-table--enable-row-transition el-table el-table--layout-fixed" data-prefix="el">
    <div class="el-table__inner-wrapper">
      <div class="hidden-columns">
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
      </div>
      <div class="el-table__header-wrapper">
        <table border="0" cellpadding="0" cellspacing="0" class="el-table__header">
          <colgroup>
            <col name="el-table_1_column_1" width="80">
            <col name="el-table_1_column_2" width="80">
            <col name="el-table_1_column_3" width="80">
            <col name="el-table_1_column_4" width="80">
            <col name="el-table_1_column_5" width="80">
            <col name="el-table_1_column_6" width="80">
          </colgroup>
          <thead class>
            <tr class>
              <th class="el-table_1_column_1 is-leaf el-table__cell" colspan="1" rowspan="1">
                <div class="cell">M_ID
                  <!---->
                  <!---->
                </div>
              </th>
              <th class="el-table_1_column_2 is-leaf el-table__cell" colspan="1" rowspan="1">
                <div class="cell">M_USERNAME
                  <!---->
                  <!---->
                </div>
              </th>
              <th class="el-table_1_column_3 is-leaf el-table__cell" colspan="1" rowspan="1">
                <div class="cell">M_CREATED_AT
                  <!---->
                  <!---->
                </div>
              </th>
              <th class="el-table_1_column_4 is-center is-leaf el-table__cell" colspan="1" rowspan="1">
                <div class="cell">M_ENABLE_STATE
                  <!---->
                  <!---->
                </div>
              </th>
              <th class="el-table_1_column_5 is-center is-leaf el-table__cell" colspan="1" rowspan="1">
                <div class="cell">M_LOCKED_STATE
                  <!---->
                  <!---->
                </div>
              </th>
              <th class="el-table_1_column_6 is-center is-leaf el-table__cell" colspan="1" rowspan="1">
                <div class="cell">M_OPERATOR
                  <!---->
                  <!---->
                </div>
              </th>
            </tr>
          </thead>
        </table>
      </div>
      <div class="el-table__body-wrapper">
        <div class="el-scrollbar">
          <div class="el-scrollbar__wrap el-scrollbar__wrap--hidden-default">
            <div class="el-scrollbar__view" style="display: inline-block; vertical-align: middle;">
              <table border="0" cellpadding="0" cellspacing="0" class="el-table__body" style="table-layout: fixed;">
                <colgroup>
                  <col name="el-table_1_column_1" width="80">
                  <col name="el-table_1_column_2" width="80">
                  <col name="el-table_1_column_3" width="80">
                  <col name="el-table_1_column_4" width="80">
                  <col name="el-table_1_column_5" width="80">
                  <col name="el-table_1_column_6" width="80">
                </colgroup>
                <!--v-if-->
                <tbody>
                  <tr class="el-table__row">
                    <td class="el-table_1_column_1 el-table__cell" colspan="1" rowspan="1">
                      <div class="cell">
                        <!---->
                      </div>
                    </td>
                    <td class="el-table_1_column_2 el-table__cell" colspan="1" rowspan="1">
                      <div class="cell">
                        <!---->user1
                      </div>
                    </td>
                    <td class="el-table_1_column_3 el-table__cell" colspan="1" rowspan="1">
                      <div class="cell">
                        <!---->2023-02-08
                      </div>
                    </td>
                    <td class="el-table_1_column_4 is-center el-table__cell" colspan="1" rowspan="1">
                      <div class="cell">
                        <!---->
                        <transition-stub appear="true" css="true" name="el-zoom-in-center" persisted="false">
                          <span class="el-tag el-tag--success el-tag--light">
                            <span class="el-tag__content">M_ENABLED</span>
                            <!--v-if-->
                          </span>
                        </transition-stub>
                      </div>
                    </td>
                    <td class="el-table_1_column_5 is-center el-table__cell" colspan="1" rowspan="1">
                      <div class="cell">
                        <!---->
                        <i class="el-icon" style="font-size: 20px;">
                          <svg height="1.2em" viewBox="0 0 1024 1024" width="1.2em">
                            <path d="M224 448a32 32 0 0 0-32 32v384a32 32 0 0 0 32 32h576a32 32 0 0 0 32-32V480a32 32 0 0 0-32-32H224zm0-64h576a96 96 0 0 1 96 96v384a96 96 0 0 1-96 96H224a96 96 0 0 1-96-96V480a96 96 0 0 1 96-96z" fill="currentColor" />
                            <path d="M512 544a32 32 0 0 1 32 32v192a32 32 0 1 1-64 0V576a32 32 0 0 1 32-32zm178.304-295.296A192.064 192.064 0 0 0 320 320v64h352l96 38.4V448H256V320a256 256 0 0 1 493.76-95.104l-59.456 23.808z" fill="currentColor" />
                          </svg>
                        </i>
                      </div>
                    </td>
                    <td class="el-table_1_column_6 is-center el-table__cell" colspan="1" rowspan="1">
                      <div class="cell">
                        <!---->
                        <button aria-disabled="false" class="el-button el-button--small" type="button">
                          <!--v-if-->
                          <span class>ACTION.EDIT</span>
                        </button>
                        <button aria-disabled="false" class="el-button el-button--danger el-button--small" type="button">
                          <!--v-if-->
                          <span class>ACTION.DELETE</span>
                        </button>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
              <!--v-if-->
              <!--v-if-->
            </div>
          </div>
          <transition-stub appear="false" css="true" name="el-scrollbar-fade" persisted="true">
            <div class="el-scrollbar__bar is-horizontal" style="display: none;">
              <div class="el-scrollbar__thumb" style="transform: translateX(0%);"></div>
            </div>
          </transition-stub>
          <transition-stub appear="false" css="true" name="el-scrollbar-fade" persisted="true">
            <div class="el-scrollbar__bar is-vertical" style="display: none;">
              <div class="el-scrollbar__thumb" style="transform: translateY(0%);"></div>
            </div>
          </transition-stub>
        </div>
      </div>
      <!--v-if-->
      <div class="el-table__border-left-patch"></div>
    </div>
    <div class="el-table__column-resize-proxy" style="display: none;"></div>
  </div>
</div>
<!--teleport start-->
<transition-stub appear="false" css="true" name="dialog-fade" persisted="true">
  <div class="el-overlay" style="z-index: 2001; display: none;">
    <div aria-describedby aria-label="M_EDITED_USER: undefined" aria-modal="true" class="el-overlay-dialog" role="dialog"></div>
  </div>
</transition-stub>
<!--teleport end-->
<!--teleport start-->
<transition-stub appear="false" css="true" name="dialog-fade" persisted="true">
  <div class="el-overlay" style="z-index: 2002; display: none;">
    <div aria-describedby aria-label="M_DELETED_USER: undefined" aria-modal="true" class="el-overlay-dialog" role="dialog"></div>
  </div>
</transition-stub>
<!--teleport end-->
`;
